apps/recnet-api/src/modules/user/user.decorator.ts

```diff
@@ -0,0 +1,10 @@
+import { createParamDecorator, ExecutionContext } from "@nestjs/common";
+
+export const User = createParamDecorator(
+  (data: string, ctx: ExecutionContext) => {
+    const request = ctx.switchToHttp().getRequest();
+    const recnetUser = request.user?.recnet;
+
+    return data ? recnetUser?.[data] : recnetUser;
+  }
+);
```

Comment:
```ts
import {
  createParamDecorator,
  ExecutionContext,
  HttpStatus,
} from "@nestjs/common";

import { RecnetError } from "@recnet-api/utils/error/recnet.error";
import { ErrorCode } from "@recnet-api/utils/error/recnet.error.const";

import { RecNetJwtPayload, recnetJwtPayloadSchema } from "@recnet/recnet-jwt";

type RecNetJwtPayloadProps = keyof RecNetJwtPayload["recnet"];

export const User = createParamDecorator<
  RecNetJwtPayloadProps | undefined,
  ExecutionContext
>((prop, ctx) => {
  const request = ctx.switchToHttp().getRequest();
  const recnetJwtPayload = recnetJwtPayloadSchema.safeParse(request.user);
  if (!recnetJwtPayload.success) {
    throw new RecnetError(
      ErrorCode.ZOD_VALIDATION_ERROR,
      HttpStatus.UNAUTHORIZED,
      "Invalid JWT payload"
    );
  }
  const recnetUser = recnetJwtPayload.data;

  return prop ? recnetUser?.[prop] : recnetUser;
});

export type AuthUser<
  Prop extends RecNetJwtPayloadProps | undefined = undefined,
> = Prop extends RecNetJwtPayloadProps
  ? RecNetJwtPayload["recnet"][Prop]
  : RecNetJwtPayload["recnet"];
```
I made my version of `user.decorator.ts` by refactoring a bit.
I would suggest these changes. Reasons:
1. Parse here for more type-safety, originally, the type of `recnetUser` is `any` . I know you've parsed at `auth.guard.ts`, I would say you can consider remove the parsing there and put it here or **just leave it there**(I prefer this). Since, the guard just a layer and do not return anything. This decorator will return value, I think run validation logic before returning value makes more sense.
2. About the type `AuthUser`. I know you've added it in `user.type.ts`. However, creating a type from scratch is not good since you lose single-source-of-truth. It has risks if we modify the RecNetJwtPayload in the future but forget to modify the `AuthUser` in another file. We can use existing type `RecNetJwtPayload` to infer `AuthUser` here.

The usage of this decorator is still the same.

You can use like this. (After the above changes, when you hover on `authUser`, you can see the type is correctly inferred still.)

```ts
public async getMe(
    @User() authUser: AuthUser
  ): Promise<GetUserMeResponse> {
    const { userId } = authUser; // the type is inferred correctly just like you did
    const user = await this.userService.getUser(userId);
    return { user };
  }
```

Or if you only want to get the "role" or "userId"

```ts
public async getMe(
    @User("role") authUserRole: AuthUser<"role">
  ): Promise<GetUserMeResponse> {
    const userRole = authUserRole; // the type here would be "ADMIN" | "USER", you can verify it
    // use this userRole to do other things...
  }

// --- or ---
public async getMe(
    @User("userId") authUserId: AuthUser<"userId">
  ): Promise<GetUserMeResponse> {
    const userId = authUserId; // the type here would be string, you can verify it
    // use this userId to do other things...
  }
```

Another advantage of creating type `AuthUser` like this is typescript will give you error if put something that's not `RecNetJwtPayloadProp` (e.g. "role", "userId") in the decorator params. In your current impl, you can write `@User("useer")` (arbitrary typos) and typescript won't give you error since the `data: string` in `createParamDecorator`